Created an announcement of Ember 3.22 release (#785)
ember-learn/ember-blog: source/2020-10-19-ember-3-22-released.md

Ember Data 3.22 introduced <!-- TODO: Add number --> deprecations.

<!-- TODO: List deprecations -->

For more details on changes in Ember Data 3.22, please review the
[Ember Data 3.22.0 release page](https://github.com/emberjs/data/releases/tag/v3.22.0).

---

## Ember CLI

Ember CLI is the command line interface for managing and packaging Ember.js applications.

### Upgrading Ember CLI

<!--alex ignore easy-->
You may upgrade Ember CLI easily using the ember-cli-update project:

```bash
npx ember-cli-update
```

This utility will help you to update your app or addon to the latest Ember CLI version. You will probably encounter merge conflicts, in which the default behavior is to let you resolve conflicts on your own. For more information on the `ember-cli-update` project, see [the GitHub README](https://github.com/ember-cli/ember-cli-update).

While it is recommended to keep Ember CLI versions in sync with Ember and Ember Data, this is not required. After updating ember-cli, you can keep your current version(s) of Ember or Ember Data by editing `package.json` to revert the changes to the lines containing `ember-source` and `ember-data`.

### Changes in Ember CLI 3.22

> #### New Features
> 
> Ember CLI 3.22 introduced <!-- TODO: Add number --> features.
> 
> <!-- TODO: List features -->
> 
> #### Deprecations
> 
> Ember CLI 3.22 introduced <!-- TODO: Add number --> deprecations.
> 
> <!-- TODO: List deprecations -->

Author: ```suggestion
Ember CLI 3.22 includes only internal changes. However, please note that [eslint-plugin-ember](https://github.com/ember-cli/eslint-plugin-ember) released a new major version, `v9.x`. The new version introduces 6 additional recommended rules and removed support of Node `v13`. To learn more, please [check out the changelog for `eslint-plugin-ember`](https://github.com/ember-cli/eslint-plugin-ember/releases/tag/v9.0.0).

#### New Features

Ember CLI 3.22 introduced 0 features.

#### Deprecations

Ember CLI 3.22 introduced 0 deprecations.
```

Today the Ember project is releasing version 3.22 of Ember.js, Ember Data, and Ember CLI.

This release kicks off the 3.23 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

You can read more about our general release process here:

- [Release Dashboard](http://emberjs.com/releases/)
- [The Ember Release Cycle](http://emberjs.com/blog/2013/09/06/new-ember-release-process.html)
- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.22

Ember.js 3.22 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations.

Ember.js 3.22 introduced several bug fixes, including:

> - [#19094](https://github.com/emberjs/ember.js/pull/19094) `RouterService#isActive` now updates tracked properties correctly by consuming `currentURL` and `currentRouteName`;
> - [#19170](https://github.com/emberjs/ember.js/pull/19170) modifier manager 3.22 now accepts the resolved value of `owner.factoryFor` directly, which is in keeping with the way helpers and component managers work; and
> - [#19124](https://github.com/emberjs/ember.js/pull/19124) `protocolForUrl` now works in fastboot `fastboot@3.1.0` environment by leveraging the native JavaScript `URL` global

Author: ```suggestion
- `RouterService#isActive` updates correctly when the route changes ([#19094](https://github.com/emberjs/ember.js/pull/19094))
- A custom modifier manager only consumes arguments that it actually uses. (Previously, the manager would eagerly consume all arguments.) In order for specific modifiers to take advantage of this change, they need to call `modifierCapabilities('3.22')` (instead of `'3.13'`). ([#19163](https://github.com/emberjs/ember.js/pull/19163))
- Modifier manager accepts the resolved value of `owner.factoryFor` directly. This approach aligns with how helpers and component managers work. ([#19170](https://github.com/emberjs/ember.js/pull/19170))
- Changes made to `protocolForUrl` (in `3.22.0-beta.2+`) work in `fastboot@3.1.0` environment by leveraging the native JavaScript `URL` global. ([#19124](https://github.com/emberjs/ember.js/pull/19124))
```

Today the Ember project is releasing version 3.22 of Ember.js, Ember Data, and Ember CLI.

This release kicks off the 3.23 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.

You can read more about our general release process here:

- [Release Dashboard](http://emberjs.com/releases/)
- [The Ember Release Cycle](http://emberjs.com/blog/2013/09/06/new-ember-release-process.html)
- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.22

Ember.js 3.22 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations.

Ember.js 3.22 introduced several bug fixes, including:

> - `RouterService#isActive` updates correctly when the route changes ([#19094](https://github.com/emberjs/ember.js/pull/19094))

Author: ```suggestion
- `RouterService#isActive` updates correctly when the route changes. ([#19094](https://github.com/emberjs/ember.js/pull/19094))
```

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.22

Ember.js 3.22 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations.

Ember.js 3.22 introduced several bug fixes, including:

- [#19094](https://github.com/emberjs/ember.js/pull/19094) `RouterService#isActive` now updates tracked properties correctly by consuming `currentURL` and `currentRouteName`;
- [#19170](https://github.com/emberjs/ember.js/pull/19170) modifier manager 3.22 now accepts the resolved value of `owner.factoryFor` directly, which is in keeping with the way helpers and component managers work; and
- [#19124](https://github.com/emberjs/ember.js/pull/19124) `protocolForUrl` now works in fastboot `fastboot@3.1.0` environment by leveraging the native JavaScript `URL` global

#### New Features

Ember.js 3.22 introduced 3 features.

- Enabled `@ember/destroyable` API, which allows you to create and manage your own destroyables. For more information, please see the [Destroyables RFC](https://emberjs.github.io/rfcs/0580-destroyables.html).

> - Enabled Cache API, a low-level primitive for memoizing the result of a function based on autotracking. (You can install a [polyfill](https://github.com/ember-polyfills/ember-cache-primitive-polyfill) for apps between versions 3.13 and 3.21.) For more information, please check the [Ember Guides](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/#toc_caching-of-tracked-properties) and [Autotracking Memoization RFC](https://emberjs.github.io/rfcs/0615-autotracking-memoization.html).

Author: ```suggestion
<!-- alex disable retext-equality -->
- Enabled Cache API, a low-level primitive for memoizing the result of a function based on autotracking. (You can install a [polyfill](https://github.com/ember-polyfills/ember-cache-primitive-polyfill) for apps between versions 3.13 and 3.21.) For more information, please check the [Ember Guides](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/#toc_caching-of-tracked-properties) and [Autotracking Memoization RFC](https://emberjs.github.io/rfcs/0615-autotracking-memoization.html).
```

- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.22

Ember.js 3.22 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations.

Ember.js 3.22 introduced several bug fixes, including:

- [#19094](https://github.com/emberjs/ember.js/pull/19094) `RouterService#isActive` now updates tracked properties correctly by consuming `currentURL` and `currentRouteName`;
- [#19170](https://github.com/emberjs/ember.js/pull/19170) modifier manager 3.22 now accepts the resolved value of `owner.factoryFor` directly, which is in keeping with the way helpers and component managers work; and
- [#19124](https://github.com/emberjs/ember.js/pull/19124) `protocolForUrl` now works in fastboot `fastboot@3.1.0` environment by leveraging the native JavaScript `URL` global

#### New Features

Ember.js 3.22 introduced 3 features.

> - Enabled `@ember/destroyable` API, which allows you to create and manage your own destroyables. For more information, please see the [Destroyables RFC](https://emberjs.github.io/rfcs/0580-destroyables.html).

Reviewer: This should be an ordered list to be consistent with the rest of the post.

Author: Thanks! I will update the post soon.

You can read more about our general release process here:

- [Release Dashboard](http://emberjs.com/releases/)
- [The Ember Release Cycle](http://emberjs.com/blog/2013/09/06/new-ember-release-process.html)
- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)

---

## Ember.js

Ember.js is the core framework for building ambitious web applications.

### Changes in Ember.js 3.22

Ember.js 3.22 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations.

Ember.js 3.22 introduced several bug fixes, including:

- [#19094](https://github.com/emberjs/ember.js/pull/19094) `RouterService#isActive` now updates tracked properties correctly by consuming `currentURL` and `currentRouteName`;

> - [#19170](https://github.com/emberjs/ember.js/pull/19170) modifier manager 3.22 now accepts the resolved value of `owner.factoryFor` directly, which is in keeping with the way helpers and component managers work; and

Reviewer 2: Not sure it matters to call out, but the 3.22 capability for modifier managers also allows modifiers to only re-run based on tracked state that they themselves consume. Previously, the framework would eagerly consume all of the arguments.

Author: Thanks for pointing this out. I think we should mention it in the blog post.

- [#19124](https://github.com/emberjs/ember.js/pull/19124) `protocolForUrl` now works in fastboot `fastboot@3.1.0` environment by leveraging the native JavaScript `URL` global

#### New Features

Ember.js 3.22 introduced 3 features.

1. Enabled `@ember/destroyable` API, which allows you to create and manage your own destroyables. For more information, please see the [Destroyables RFC](https://emberjs.github.io/rfcs/0580-destroyables.html).

> 2. Enabled Cache API, a low-level primitive for memoizing the result of a function based on autotracking. (You can install a [polyfill](https://github.com/ember-polyfills/ember-cache-primitive-polyfill) for apps between versions 3.13 and 3.21.) For more information, please check the [Ember Guides](https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/#toc_caching-of-tracked-properties) and [Autotracking Memoization RFC](https://emberjs.github.io/rfcs/0615-autotracking-memoization.html).

Reviewer 2: Is it possible to link to the published API docs we have for these? It would be pretty nice if we could...

3. Allowed passing the `transition` object to 2 route hooks, `activate` and `deactivate`.

#### Deprecations

Ember.js 3.22 introduced 0 deprecations.

Deprecations are added to Ember.js when an API will be removed at a later date. Each deprecation has an entry in the deprecation guide describing the migration path to a more stable API. Deprecated public APIs are not removed until a major release of the framework.

Consider using the [ember-cli-deprecation-workflow](https://github.com/mixonic/ember-cli-deprecation-workflow) addon if you would like to upgrade your application without immediately addressing deprecations.

> For more details on changes in Ember.js 3.22, please review the [Ember.js 3.22.0 release page](https://github.com/emberjs/ember.js/releases/tag/v3.22.0).

Reviewer 2: I think I'd probably remove this bit if we didn't introduce any depreciations no need to tell them about the deprecation workflow...

Author: Sounds good with me, thanks for your suggestion. Yeah, the paragraphs were a copy-paste from a template.

```suggestion
```

For more details on changes in Ember Data 3.22, please review the
[Ember Data 3.22.0 release page](https://github.com/emberjs/data/releases/tag/v3.22.0).

---

## Ember CLI

Ember CLI is the command line interface for managing and packaging Ember.js applications.

### Upgrading Ember CLI

<!--alex ignore easy-->
You may upgrade Ember CLI easily using the ember-cli-update project:

```bash
npx ember-cli-update
```

This utility will help you to update your app or addon to the latest Ember CLI version. You will probably encounter merge conflicts, in which the default behavior is to let you resolve conflicts on your own. For more information on the `ember-cli-update` project, see [the GitHub README](https://github.com/ember-cli/ember-cli-update).

While it is recommended to keep Ember CLI versions in sync with Ember and Ember Data, this is not required. After updating ember-cli, you can keep your current version(s) of Ember or Ember Data by editing `package.json` to revert the changes to the lines containing `ember-source` and `ember-data`.

### Changes in Ember CLI 3.22

> Ember CLI 3.22 includes only internal changes. However, please note that [eslint-plugin-ember](https://github.com/ember-cli/eslint-plugin-ember) released a new major version, `v9.x`. The new version introduces 6 additional recommended rules and removed support of Node `v13`. To learn more, please [check out the changelog for `eslint-plugin-ember`](https://github.com/ember-cli/eslint-plugin-ember/releases/tag/v9.0.0).
> 

Author: ```suggestion
Ember CLI 3.22 includes only internal changes. However, please note that [eslint-plugin-ember](https://github.com/ember-cli/eslint-plugin-ember) released a new major version, `v9.x`. The new version introduces 6 recommended rules and removes support of Node `v13`.

New Ember apps will install `v9` or higher of `eslint-plugin-ember`, which may limit the version of Node that you use. To learn more, please [check out the changelog for `eslint-plugin-ember`](https://github.com/ember-cli/eslint-plugin-ember/releases/tag/v9.0.0).
```
